File: desktop/crates/core/src/catalog.rs
//! The phone's listing of the files it holds.

use crate::id::{DevicePath, Timestamp};

/// One file as the phone sees it. The three fields together are the identity used by the diff,
/// the staging manifest, and the deletion candidates.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct CatalogEntry {
    pub path: DevicePath,
    pub size: u64,
    pub mtime: Timestamp,
}

/// A catalog frozen at the start of a session. It never changes while the session is alive,
/// including across reconnections, so a repeated diff can only shrink the work.
#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct Catalog {
    entries: Vec<CatalogEntry>,
    total_bytes: u64,
}

impl Catalog {
    pub fn new(entries: Vec<CatalogEntry>, total_bytes: u64) -> Self {
        Self {
            entries,
            total_bytes,
        }
    }

    pub fn entries(&self) -> &[CatalogEntry] {
        &self.entries
    }

    pub fn total_bytes(&self) -> u64 {
        self.total_bytes
    }

    pub fn is_empty(&self) -> bool {
        self.entries.is_empty()
    }
}

File: android/session/src/main/kotlin/top/wxx9248/photosync/session/Catalog.kt
package top.wxx9248.photosync.session

/**
 * One photograph as the phone sees it.
 *
 * `SPEC.md` §3.2 fixes the identity of an entry as the three fields together, and everything
 * else in the protocol rests on that: the diff matches on it, the desktop refuses an upload
 * whose header disagrees with it, and the cheap deletion gate compares it.
 */
data class CatalogEntry(
    val path: DevicePath,
    val size: Long,
    val mtime: Timestamp,
)

/**
 * Everything the phone is offering, frozen.
 *
 * §3.2 enumerates once, at the start of a session, and holds the result for the whole of it.
 * A reconnection re-sends this same list, which is what makes a re-diff able only to shrink
 * the work. Freezing it is the reason this type exists rather than a bare list: a caller
 * cannot add to it half way through.
 */
class Catalog(entries: List<CatalogEntry>) {
    val entries: List<CatalogEntry> = entries.toList()

    val totalBytes: Long = entries.sumOf { it.size }

    operator fun get(path: DevicePath): CatalogEntry? = entries.firstOrNull { it.path == path }
}
